src/circuit/circuit_frame.rs

```diff
@@ -1207,21 +1239,34 @@ fn reduce_cons<F: PrimeField, CS: ConstraintSystem<F>>(
         let bindings_is_nil =
             bindings.alloc_equal(&mut cs_letrec.namespace(|| "bindings_is_nil"), &g.nil_ptr)?;
 
-        let (val1, _end) = car_cdr(
+        let rest_body_is_nil =
+            rest_body.alloc_equal(&mut cs_letrec.namespace(|| "rest_body_is_nil"), &g.nil_ptr)?;
+
+        let (val, end) = car_cdr(
             &mut cs_letrec.namespace(|| "car_cdr more_vals"),
             g,
-            &more_vals,
+            &vals,
             store,
         )?;
 
-        let val = AllocatedPtr::pick(
-            &mut cs_letrec.namespace(|| "pick val"),
-            &bindings_is_nil,
-            &body1,
-            &val1,
+        let end_is_nil = end.alloc_equal(&mut cs_letrec.namespace(|| "end_is_nil"), &g.nil_ptr)?;
+
+        /*
+         * We get the condition for error by using OR of each individual error.
+         */
+        // First error happens when binding is not nil AND end is not nil
+        let mut cond_error = Boolean::and(
+            &mut cs_letrec.namespace(|| "if binding not nil and end is nil"),
+            &Boolean::not(&bindings_is_nil),
+            &Boolean::not(&end_is_nil),
+        )?;
+        // Second error happens when rest_body is not nil
+        cond_error = constraints::or(
+            &mut cs_letrec.namespace(|| "cond error"),
+            &Boolean::not(&rest_body_is_nil),
+            &cond_error,
```

Comment:
Taken in conjunction with my comment above, I think the high-level meaning of this condition is meant to be: "Either there is more than one body element, or there is more than one value in a binding."

I agree that both of these are error conditions, so maybe we can just substitute the `not(&end_is_nil)` for `cond_error` here.